File: backend/schemas/theme_setup.py
"""테마 셋업 스키마."""
from datetime import date, datetime
from typing import Optional
from pydantic import BaseModel


class TopStockItem(BaseModel):
    """상위 패턴 종목."""
    code: str
    name: str
    pattern: str
    confidence: int


class ScoreBreakdownNews(BaseModel):
    """뉴스 점수 상세."""
    score: float
    count_7d: Optional[int] = None
    wow_change: Optional[int] = None
    source_diversity: Optional[float] = None


class ScoreBreakdownChart(BaseModel):
    """차트 패턴 점수 상세."""
    score: float
    pattern_ratio: Optional[float] = None
    avg_confidence: Optional[float] = None
    patterns: Optional[list[str]] = None
    pattern_count: Optional[int] = None
    total_stocks: Optional[int] = None


class ScoreBreakdownMention(BaseModel):
    """언급 점수 상세."""
    score: float
    youtube_count: Optional[int] = None
    trader_count: Optional[int] = None


class ScoreBreakdownPrice(BaseModel):
    """가격 액션 점수 상세."""
    score: float
    avg_change: Optional[float] = None
    volume_change: Optional[float] = None


class ScoreBreakdownFlow(BaseModel):
    """수급 점수 상세."""
    score: float
    foreign_net_sum: Optional[int] = None
    institution_net_sum: Optional[int] = None
    positive_foreign: Optional[int] = None
    positive_institution: Optional[int] = None
    total_stocks: Optional[int] = None
    avg_flow_score: Optional[float] = None


class ScoreBreakdown(BaseModel):
    """점수 상세 breakdown."""
    news: Optional[dict] = None
    chart: Optional[dict] = None
    mention: Optional[dict] = None
    price: Optional[dict] = None
    flow: Optional[dict] = None


class ThemeSetupResponse(BaseModel):
    """테마 셋업 응답."""
    theme_name: str
    rank: int
    total_score: float
    news_momentum_score: float
    chart_pattern_score: float
    mention_score: float
    price_action_score: float
    investor_flow_score: float = 0.0
    top_stocks: list[dict]
    stocks_with_pattern: int
    total_stocks: int
    is_emerging: int
    score_breakdown: Optional[dict] = None
    explanation: Optional[str] = None  # 점수 산출 이유 한글 설명


class ThemeSetupDetailResponse(BaseModel):
    """테마 셋업 상세 응답."""
    theme_name: str
    rank: int
    total_score: float
    news_momentum_score: float
    chart_pattern_score: float
    mention_score: float
    price_action_score: float
    investor_flow_score: float = 0.0
    score_breakdown: Optional[dict] = None
    top_stocks: list[dict]
    stocks_with_pattern: int
    total_stocks: int
    is_emerging: int
    setup_date: str
    history: list[dict]


class ChartPatternResponse(BaseModel):
    """차트 패턴 응답."""
    stock_code: str
    stock_name: str
    pattern_type: str
    confidence: int
    pattern_data: dict
    current_price: int
    price_from_support_pct: Optional[float] = None
    price_from_resistance_pct: Optional[float] = None


class NewsTrendResponse(BaseModel):
    """뉴스 추이 응답."""
    date: str
    mention_count: int
    unique_sources: int
    top_keywords: list[dict]
    wow_change_pct: Optional[int] = None


class EmergingThemesResponse(BaseModel):
    """이머징 테마 목록 응답."""
    themes: list[ThemeSetupResponse]
    total_count: int
    generated_at: str


class ThemeSetupCalculateRequest(BaseModel):
    """셋업 계산 요청."""
    theme_names: Optional[list[str]] = None  # None이면 전체


class ThemeSetupCalculateResponse(BaseModel):
    """셋업 계산 응답."""
    calculated_count: int
    emerging_count: int
    timestamp: str

File: frontend/src/types/theme_setup.ts
// Theme Setup Types - 테마 셋업 (자리 잡는 테마 감지)

export interface TopPatternStock {
  code: string
  name: string
  pattern: string
  confidence: number
}

export interface ThemeSetup {
  theme_name: string
  rank: number
  total_score: number
  news_momentum_score: number
  chart_pattern_score: number
  mention_score: number
  price_action_score: number
  investor_flow_score: number
  top_stocks: TopPatternStock[]
  stocks_with_pattern: number
  total_stocks: number
  is_emerging: number
  explanation?: string  // 점수 산출 이유 설명
  score_breakdown?: ScoreBreakdown
}

export interface ScoreBreakdown {
  news: {
    score: number
    '7d_count'?: number
    wow_change?: number
    source_diversity?: number
  }
  chart: {
    score: number
    pattern_ratio?: number
    avg_confidence?: number
    patterns?: string[]
    pattern_count?: number
    total_stocks?: number
  }
  mention: {
    score: number
    youtube_count?: number
    expert_count?: number
  }
  price: {
    score: number
    avg_change?: number
    volume_change?: number
  }
  flow?: {
    score: number
    foreign_net_sum?: number
    institution_net_sum?: number
    positive_foreign?: number
    positive_institution?: number
    total_stocks?: number
    avg_flow_score?: number
  }
}

export interface ThemeSetupHistory {
  date: string
  score: number
  rank: number
}

export interface ThemeSetupDetail extends ThemeSetup {
  score_breakdown: ScoreBreakdown
  setup_date: string
  history: ThemeSetupHistory[]
}

export interface ChartPattern {
  stock_code: string
  stock_name: string
  pattern_type: string
  confidence: number
  pattern_data: Record<string, unknown>
  current_price: number
  price_from_support_pct?: number
  price_from_resistance_pct?: number
}

export interface NewsTrendItem {
  date: string
  mention_count: number
  unique_sources: number
  top_keywords: Array<{ keyword: string; count: number }>
  wow_change_pct?: number
}

export interface ThemeNewsItem {
  title: string
  url: string
  source: string
  published_at: string
  keyword: string
}

export interface EmergingThemesResponse {
  themes: ThemeSetup[]
  total_count: number
  generated_at: string
}

export interface SetupHistoryItem {
  date: string
  total_score: number
  news_score: number
  chart_score: number
  mention_score: number
  price_score: number
  flow_score: number
  rank: number
}

// 패턴 타입 한글 매핑
export const PATTERN_TYPE_LABELS: Record<string, string> = {
  range_bound: '횡보/박스권',
  double_bottom: '쌍바닥',
  triple_bottom: '삼중바닥',
  converging: '수렴',
  pre_breakout: '돌파 직전',
}

// 패턴별 색상
export const PATTERN_TYPE_COLORS: Record<string, string> = {
  range_bound: 'bg-gray-100 text-gray-800',
  double_bottom: 'bg-green-100 text-green-800',
  triple_bottom: 'bg-emerald-100 text-emerald-800',
  converging: 'bg-blue-100 text-blue-800',
  pre_breakout: 'bg-orange-100 text-orange-800',
}
